File: src/agentic_tools/main.test.mjs
// @ts-check

import { describe, expect, test } from "@jest/globals";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

import { runAgenticTools } from "./main.mjs";

const createTempDir = () => {
  return fs.mkdtempSync(path.join(os.tmpdir(), "agentic-tools-node-root-"));
};

/** @param {string} tempDir */
const cleanupTempDir = (tempDir) => {
  fs.rmSync(tempDir, { recursive: true, force: true });
};

describe("agentic-tools Node CLI", () => {
  test("runAgenticTools dispatches policy sync", async () => {
    const tempDir = createTempDir();
    try {
      fs.mkdirSync(path.join(tempDir, ".agents"), { recursive: true });
      fs.writeFileSync(
        path.join(tempDir, ".agents", "policy.json"),
        JSON.stringify(
          {
            services: ["copilot"],
            protectedFiles: ["*.env"],
          },
          null,
          2,
        ),
        "utf8",
      );

      const exitCode = await runAgenticTools(["policy", "sync"], {
        cwd: tempDir,
        output: () => {},
      });

      const vscodeSettings = /** @type {Record<string, Record<string, string>>} */ (JSON.parse(
        fs.readFileSync(path.join(tempDir, ".vscode", "settings.json"), "utf8"),
      ));

      expect(exitCode).toBe(0);
      expect(vscodeSettings["files.associations"]).toEqual({
        "*.env": "copilot-restricted-file",
      });
    } finally {
      cleanupTempDir(tempDir);
    }
  });

  test("runAgenticTools dispatches policy import-vscode", async () => {
    const tempDir = createTempDir();
    try {
      fs.mkdirSync(path.join(tempDir, ".agents"), { recursive: true });
      fs.mkdirSync(path.join(tempDir, ".vscode"), { recursive: true });
      fs.writeFileSync(
        path.join(tempDir, ".agents", "policy.json"),
        JSON.stringify(
          {
            services: ["copilot"],
            terminalAutoApprove: { stale: true },
          },
          null,
          2,
        ),
        "utf8",
      );
      fs.writeFileSync(
        path.join(tempDir, ".vscode", "settings.json"),
        JSON.stringify(
          {
            "chat.tools.terminal.autoApprove": { "uv run poe test": true },
          },
          null,
          2,
        ),
        "utf8",
      );

      const exitCode = await runAgenticTools(["policy", "import-vscode"], {
        cwd: tempDir,
        output: () => {},
      });

      const policy = /** @type {Record<string, unknown>} */ (JSON.parse(
        fs.readFileSync(path.join(tempDir, ".agents", "policy.json"), "utf8"),
      ));

      expect(exitCode).toBe(0);
      expect(policy.terminalAutoApprove).toEqual({ "uv run poe test": true });
    } finally {
      cleanupTempDir(tempDir);
    }
  });

  test("runAgenticTools dispatches skills list", async () => {
    const tempDir = createTempDir();
    try {
      const skillDir = path.join(tempDir, ".agents", "skills", "ref-alpha");
      fs.mkdirSync(skillDir, { recursive: true });
      fs.writeFileSync(
        path.join(skillDir, "SKILL.md"),
        [
          "---",
          "name: ref-alpha",
          'description: "Test skill."',
          "metadata:",
          '  shareable-skills.visibility: "shareable"',
          "---",
          "",
          "# Test Skill",
          "",
        ].join("\n"),
        "utf8",
      );

      /** @type {string[]} */
      const messages = [];
      const exitCode = await runAgenticTools(["skills", "list", "--from", tempDir], {
        cwd: tempDir,
        output: (message) => {
          messages.push(message);
        },
      });

      expect(exitCode).toBe(0);
      expect(messages.join("\n")).toMatch(/ref-alpha/u);
    } finally {
      cleanupTempDir(tempDir);
    }
  });

  test("runAgenticTools without arguments prints root usage", async () => {
    /** @type {string[]} */
    const messages = [];

    const exitCode = await runAgenticTools([], {
      output: (message) => {
        messages.push(message);
      },
    });

    expect(exitCode).toBe(1);
    expect(messages.join("\n")).toMatch(/agentic-tools policy\|skills/u);
    expect(messages.join("\n")).toContain("Sync, check, and import generated agent policy files.");
    expect(messages.join("\n")).toContain("List, link, sync, and unlink shareable skills.");
  });
});

File: src/agentic_tools_old/main_test.py
import json
from pathlib import Path

import pytest

from agentic_tools_old.main import main


def write_json(path: Path, data: object) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")


def read_json(path: Path) -> dict[str, object]:
    return json.loads(path.read_text(encoding="utf-8"))


def test_main_dispatches_policy_sync(tmp_path: Path) -> None:
    repo_root = tmp_path / "repo"
    policy_path = repo_root / ".agents" / "policy.json"
    write_json(
        policy_path,
        {
            "services": ["copilot"],
            "protectedFiles": ["*.env"],
        },
    )

    exit_code = main(["policy", "sync", "--config", str(policy_path)])

    assert exit_code == 0
    vscode_settings = read_json(repo_root / ".vscode" / "settings.json")
    assert vscode_settings["files.associations"] == {"*.env": "copilot-restricted-file"}


def test_main_dispatches_policy_sync_with_workspace(tmp_path: Path) -> None:
    repo_root = tmp_path / "repo"
    policy_path = repo_root / ".agents" / "policy.json"
    write_json(
        policy_path,
        {
            "services": ["copilot"],
            "protectedFiles": ["*.env"],
        },
    )

    exit_code = main(["--workspace", str(repo_root), "policy", "sync"])

    assert exit_code == 0
    vscode_settings = read_json(repo_root / ".vscode" / "settings.json")
    assert vscode_settings["files.associations"] == {"*.env": "copilot-restricted-file"}


def test_main_dispatches_policy_import_vscode(tmp_path: Path) -> None:
    repo_root = tmp_path / "repo"
    policy_path = repo_root / ".agents" / "policy.json"
    write_json(
        policy_path,
        {
            "services": ["copilot"],
            "terminalAutoApprove": {"stale": True},
        },
    )
    write_json(
        repo_root / ".vscode" / "settings.json",
        {
            "chat.tools.terminal.autoApprove": {"/^uv run poe test$/": True},
        },
    )

    exit_code = main(["policy", "import-vscode", "--config", str(policy_path)])

    assert exit_code == 0
    policy = read_json(policy_path)
    assert policy["terminalAutoApprove"] == {"/^uv run poe test$/": True}


def test_main_dispatches_skills_list(
    tmp_path: Path, capsys: pytest.CaptureFixture[str]
) -> None:
    skills_root = tmp_path / ".agents" / "skills" / "ref-alpha"
    skills_root.mkdir(parents=True)
    (skills_root / "SKILL.md").write_text(
        "\n".join(
            [
                "---",
                "name: ref-alpha",
                'description: "Test skill."',
                "metadata:",
                '  shareable-skills.visibility: "shareable"',
                "---",
                "",
                "# Test Skill",
                "",
            ]
        ),
        encoding="utf-8",
    )

    exit_code = main(["skills", "list", "--from", str(tmp_path)])

    assert exit_code == 0
    assert "ref-alpha" in capsys.readouterr().out


def test_main_dispatches_skills_list_with_workspace(
    tmp_path: Path, capsys: pytest.CaptureFixture[str]
) -> None:
    repo_root = tmp_path / "repo"
    skills_root = repo_root / ".agents" / "skills" / "ref-alpha"
    skills_root.mkdir(parents=True)
    (skills_root / "SKILL.md").write_text(
        "\n".join(
            [
                "---",
                "name: ref-alpha",
                'description: "Test skill."',
                "metadata:",
                '  shareable-skills.visibility: "shareable"',
                "---",
                "",
                "# Test Skill",
                "",
            ]
        ),
        encoding="utf-8",
    )

    exit_code = main(["--workspace", str(repo_root), "skills", "list"])

    assert exit_code == 0
    assert "ref-alpha" in capsys.readouterr().out


def test_main_without_arguments_prints_help(capsys: pytest.CaptureFixture[str]) -> None:
    exit_code = main([])

    assert exit_code == 1
    assert "policy" in capsys.readouterr().out
